### PR TITLE
fix(mpack): correct boundary values for negative integer encoding

### LINKED ISSUE #37202: vim.mpack turns -129 to 127
### Problem

When decoding after encoding -129 using `vim.mpack`, it turns into 127.

### Steps to reproduce

```
nvim --clean
:=vim.mpack.decode(vim.mpack.encode(-129))
```

### Expected behavior

Not turning -129 to 127.

### Nvim version (nvim -v)

v0.12.0-nightly+e916f03

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Linux nixos

### Terminal name/version

kitty 0.45.0

### $TERM environment variable

xterm-kitty

### Installation

system package manager

### PR DESCRIPTION
**Pull request by benarcher2691**
## Summary

- Fix off-by-one errors in mpack encoding for negative integer boundary values
- -129 was incorrectly encoded as int8 (decoding to 127) instead of int16
- -32769 was incorrectly encoded as int16 instead of int32

## Root Cause

The boundary checks in `mpack_pack_number` (conv.c) and `mpack_wnint` (mpack_core.c) used exclusive comparisons (`<`) with the exact boundary values:

```c
// Before (wrong):
else if (tok.data.value.lo < 0xffffff7f) tok.length = 2;  // -129 in two's complement

// After (correct):
else if (tok.data.value.lo < 0xffffff80) tok.length = 2;  // -128 in two's complement
```

Since -129 equals `0xffffff7f` in two's complement, the condition `lo < 0xffffff7f` was FALSE, causing it to fall through to int8 encoding instead of int8.

## Test plan

- [x] Added test for boundary values (-128, -129, -32768, -32769)
- [x] `make lintc` passes
- [x] `TEST_FILE=test/functional/lua/mpack_spec.lua make functionaltest` passes

Fixes #37202

### CODE REVIEW

**Comment by wsdjeg** [comment]
Is it ok to merge code co-authored-by AI?

**Comment by benarcher2691** [comment]
@wsdjeg
I really don't know. If not, I am in deep waters :-) Let's await response from neovim crew.

**Comment by clason** [comment]
We don't care _how_ you wrote the code, but we do care that you fully(!) understand it and how it solves the underlying issue. LLM coding assistants can help with tedious routine and investigation (such as constructing test cases), but they are _not_ a replacement for understanding the problem as well as the code you touch. (Nvim's codebase is full of... let's say "history", and generic models tend to do quite poorly here.)

What is _not_ OK is to copy-paste responses from the LLM as your comments. We don't want to play a game of telephone with the LLM (if it was smart enough to solve the problem, we would be doing that ourselves).

Except in special circumstances (and with explicit notes), all your comments and descriptions must be written by you yourself. (Use a translation tool if you must, but don't let someone else put words in your mouth.)

**Comment by clason** [comment]
Also, `libmpack` is a vendored dependency. Please contribute the fix upstream to https://github.com/libmpack/libmpack, and we'll sync with whatever they merge (which tends to be the PR as-is).

**Comment by benarcher2691** [comment]
Got it. I close this PR since it is too heavy leaning on Claude Code. Sorry for any inconvenience. All the best!

**Comment by clason** [comment]
As I said, we don't care where the code comes from as long as it's good (and conversations happen between people, not software).

**Comment by justinmk** [comment]
cc @tarruda

**Review by tarruda — APPROVED** [review]
LGTM

Will also push to https://github.com/libmpack/libmpack

thanks @justinmk

File: src/mpack/conv.c [hunk 1]
```diff
@@ -122,8 +122,8 @@ MPACK_API mpack_token_t mpack_pack_number(double v)
     if (!tok.data.value.lo) tok.data.value.hi++;
     if (tok.data.value.lo == 0 && tok.data.value.hi == 0) tok.length = 1;
     else if (tok.data.value.lo < 0x80000000) tok.length = 8;
-    else if (tok.data.value.lo < 0xffff7fff) tok.length = 4;
-    else if (tok.data.value.lo < 0xffffff7f) tok.length = 2;
+    else if (tok.data.value.lo < 0xffff8000) tok.length = 4;
+    else if (tok.data.value.lo < 0xffffff80) tok.length = 2;
```
Comment (justinmk) [inline src/mpack/conv.c:126]: The functional test didn't cover this conv.c change so I added a unittest.
